clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -290,17 +315,47 @@ public void poll(long timeout) throws IOException {
 
         clear();
 
-        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty())
+        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty() || !keysWithBytesBuffered.isEmpty())
             timeout = 0;
 
+        boolean poolOutOfMemory = memoryPool.isOutOfMemory();
+        if (poolOutOfMemory && !outOfMemory) {
+            //dont bother selecting for reads
+            outOfMemory = true;
+            log.trace("Broker is low on memory - muting incoming sockets");
+            for (KafkaChannel channel : channels.values()) {
+                if (channel.isInMutableState())
+                    channel.mute();
+            }
```

Comment:
I am not sure of the value of this loop. It is muting a subset of channels (ones that are not in handshake and have not allocated memory and have started read). Channels not muted here and new channels are muted when and only when allocation for read fails. Wouldn't it be better to do the same for the subset handled here as well and remove this loop altogether? It seems to me that this loop simply prevents channels from reading the 4-byte size for which space has already been allocated.